[SPARK-10745][Core]Separate configs between shuffle and RPC

zsxwing opened the pull request:
[SPARK-6028](https://issues.apache.org/jira/browse/SPARK-6028) uses network module to implement RPC. However, there are some configurations named with `spark.shuffle` prefix in the network module. 

This PR refactors them to make sure the user can control them in shuffle and RPC separately. The user can use `spark.rpc.*` to set the configuration for netty RPC.

zsxwing committed:
Separate configs between shuffle and RPC

zsxwing commented:
/cc @rxin @vanzin

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45079 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45079/consoleFull)** for PR 9481 at commit [`c0c8973`](https://github.com/apache/spark/commit/c0c897389b8375ff2ef9d3b52605a0b6c1acd94e).

SparkQA commented:
**[Test build #45079 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45079/consoleFull)** for PR 9481 at commit [`c0c8973`](https://github.com/apache/spark/commit/c0c897389b8375ff2ef9d3b52605a0b6c1acd94e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class PrefixSpanModel(JavaModelWrapper):`\n  * `class PrefixSpan(object):`\n  * `    class FreqSequence(namedtuple(\"FreqSequence\", [\"sequence\", \"freq\"])):`\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45079/
Test PASSed.

vanzin commented:
I was hoping that you'd change the network library to now use "shuffle" in the config names, since it's not used only for shuffle anymore... and make that some sort of parameter to the conf objects instead.

zsxwing committed:
Add network configurations and move shuffle configurations to ShuffleConfigProvider

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

zsxwing commented:
I renamed `spark.shuffle.*` configurations to `spark.network.*` and added a `ShuffleConfigProvider` to convert them.

Now the strategy is as follows:

1. For shuffle TransportConf, it will use `spark.shuffle.*` at first, if not found, use `spark.network.*` instead.
2. For RPC TransportConf, it will use `spark.rpc.*` at first, if not found, use `spark.network.*` instead.

In other words, the user can use `spark.network.*` to set configurations for both shuffle and RPC, and also can use  `spark.rpc.*` or  `spark.shuffle.*` to set separately.

SparkQA commented:
**[Test build #45159 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45159/consoleFull)** for PR 9481 at commit [`feb5962`](https://github.com/apache/spark/commit/feb59621fffa5240e7e7fc74752d3707c50bf7a3).

SparkQA commented:
**[Test build #45159 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45159/consoleFull)** for PR 9481 at commit [`feb5962`](https://github.com/apache/spark/commit/feb59621fffa5240e7e7fc74752d3707c50bf7a3).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `public class ShuffleConfigProvider extends ConfigProvider `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45159/
Test FAILed.

zsxwing committed:
Add license

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45164 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45164/consoleFull)** for PR 9481 at commit [`2b9dff3`](https://github.com/apache/spark/commit/2b9dff361ceebd327319c5354c77f1b19f271329).

vanzin commented:
This looks very complicated and error prone. Why not have `TransportConf` take a `module` parameter:

    TransportConf(String module, ConfigProvider ...)

Then instantiate it like this:

    new TransportConf("shuffle", ...)
    new TransportConf("rpc", ...)

And have `TransportConf` insert `module` into the conf key when retrieving values?

SparkQA commented:
**[Test build #45164 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45164/consoleFull)** for PR 9481 at commit [`2b9dff3`](https://github.com/apache/spark/commit/2b9dff361ceebd327319c5354c77f1b19f271329).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `public class ShuffleConfigProvider extends ConfigProvider `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45164/
Test FAILed.

vanzin commented:
@zsxwing do you plan to update this? We should get some version of this change in 1.6 if the netty rpc layer is left as the default.

zsxwing commented:
> @zsxwing do you plan to update this? We should get some version of this change in 1.6 if the netty rpc layer is left as the default.

Agree. Just working on other stuff. But should be able to update this PR today.

zsxwing committed:
Add 'module' paramter to TransportConf

SparkQA commented:
**[Test build #46117 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46117/consoleFull)** for PR 9481 at commit [`ac81ce5`](https://github.com/apache/spark/commit/ac81ce51e015e2c4302c5c25d9e95e72caad234b).

zsxwing committed:
Add param docs

SparkQA commented:
**[Test build #46119 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46119/consoleFull)** for PR 9481 at commit [`2714294`](https://github.com/apache/spark/commit/2714294340555151ad9e2de19d46d9cb584cc523).

zsxwing commented:
@vanzin updated as per your comments.

SparkQA commented:
**[Test build #46117 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46117/consoleFull)** for PR 9481 at commit [`ac81ce5`](https://github.com/apache/spark/commit/ac81ce51e015e2c4302c5c25d9e95e72caad234b).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46117/
Test FAILed.

zsxwing committed:
Fix 'spark.shuffle.io.lazyFD' as well

SparkQA commented:
**[Test build #46122 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46122/consoleFull)** for PR 9481 at commit [`eb411bc`](https://github.com/apache/spark/commit/eb411bc0b47d4a2837c02b6e76eaa19dff5cc72d).

SparkQA commented:
**[Test build #46119 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46119/consoleFull)** for PR 9481 at commit [`2714294`](https://github.com/apache/spark/commit/2714294340555151ad9e2de19d46d9cb584cc523).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46119/
Test FAILed.

zsxwing commented:
retest this please

SparkQA commented:
**[Test build #46128 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46128/consoleFull)** for PR 9481 at commit [`eb411bc`](https://github.com/apache/spark/commit/eb411bc0b47d4a2837c02b6e76eaa19dff5cc72d).

SparkQA commented:
**[Test build #46122 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46122/consoleFull)** for PR 9481 at commit [`eb411bc`](https://github.com/apache/spark/commit/eb411bc0b47d4a2837c02b6e76eaa19dff5cc72d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46122/
Test FAILed.

SparkQA commented:
**[Test build #46128 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46128/consoleFull)** for PR 9481 at commit [`eb411bc`](https://github.com/apache/spark/commit/eb411bc0b47d4a2837c02b6e76eaa19dff5cc72d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46128/
Test FAILed.

zsxwing commented:
retest this please

zsxwing committed:
Fix tests

SparkQA commented:
**[Test build #46156 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46156/consoleFull)** for PR 9481 at commit [`523decb`](https://github.com/apache/spark/commit/523decbfb3cb47a10c0c45f0c87432e9a4100983).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46153/
Test FAILed.

zsxwing commented:
retest this please

SparkQA commented:
**[Test build #46160 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46160/consoleFull)** for PR 9481 at commit [`523decb`](https://github.com/apache/spark/commit/523decbfb3cb47a10c0c45f0c87432e9a4100983).

vanzin commented on core/src/main/scala/org/apache/spark/network/netty/SparkTransportConf.scala:
```diff
@@ -40,23 +40,26 @@ object SparkTransportConf {
    */
-  def fromSparkConf(_conf: SparkConf, numUsableCores: Int = 0): TransportConf = {
+  def fromSparkConf(_conf: SparkConf, module: String, numUsableCores: Int = 0):
+    TransportConf = {
```
this fits in the previous line, no?

vanzin commented on core/src/main/scala/org/apache/spark/network/netty/SparkTransportConf.scala:
```diff
@@ -40,23 +40,26 @@ object SparkTransportConf {
-      conf.get("spark.shuffle.io.serverThreads", numThreads.toString))
-    conf.set("spark.shuffle.io.clientThreads",
-      conf.get("spark.shuffle.io.clientThreads", numThreads.toString))
+    conf.set("spark." + module + ".io.serverThreads",
```
Not your fault, but this could be `setIfMissing`, no? Also, interpolation would look a little better.

vanzin commented:
LGTM, just nits left.

zsxwing committed:
Address comments

SparkQA commented:
**[Test build #46172 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46172/consoleFull)** for PR 9481 at commit [`3fd22e1`](https://github.com/apache/spark/commit/3fd22e14a03acddf87ec6985700e1dfffb11d1c4).

SparkQA commented:
**[Test build #46156 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46156/consoleFull)** for PR 9481 at commit [`523decb`](https://github.com/apache/spark/commit/523decbfb3cb47a10c0c45f0c87432e9a4100983).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46156/
Test FAILed.

SparkQA commented:
**[Test build #46160 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46160/consoleFull)** for PR 9481 at commit [`523decb`](https://github.com/apache/spark/commit/523decbfb3cb47a10c0c45f0c87432e9a4100983).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46160/
Test FAILed.

SparkQA commented:
**[Test build #46172 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46172/consoleFull)** for PR 9481 at commit [`3fd22e1`](https://github.com/apache/spark/commit/3fd22e14a03acddf87ec6985700e1dfffb11d1c4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46172/
Test FAILed.

rxin commented:
What's going on with jenkins?

SparkQA commented:
**[Test build #2084 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2084/consoleFull)** for PR 9481 at commit [`3fd22e1`](https://github.com/apache/spark/commit/3fd22e14a03acddf87ec6985700e1dfffb11d1c4).

zsxwing committed:
Merge remote-tracking branch 'origin/master' into SPARK-10745

zsxwing committed:
Fix tests

zsxwing commented:
> What's going on with jenkins?

Forgot to merge master. Some new commits to master needs to be updated as well.

SparkQA commented:
**[Test build #46185 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46185/consoleFull)** for PR 9481 at commit [`84b7a42`](https://github.com/apache/spark/commit/84b7a4278771efb9c0ce158afef3620535eada07).

SparkQA commented:
**[Test build #2084 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2084/consoleFull)** for PR 9481 at commit [`3fd22e1`](https://github.com/apache/spark/commit/3fd22e14a03acddf87ec6985700e1dfffb11d1c4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #46185 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46185/consoleFull)** for PR 9481 at commit [`84b7a42`](https://github.com/apache/spark/commit/84b7a4278771efb9c0ce158afef3620535eada07).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46185/
Test FAILed.

zsxwing commented:
retest this please

SparkQA commented:
**[Test build #46223 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46223/consoleFull)** for PR 9481 at commit [`84b7a42`](https://github.com/apache/spark/commit/84b7a4278771efb9c0ce158afef3620535eada07).

SparkQA commented:
**[Test build #46223 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46223/consoleFull)** for PR 9481 at commit [`84b7a42`](https://github.com/apache/spark/commit/84b7a4278771efb9c0ce158afef3620535eada07).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46223/
Test PASSed.

vanzin commented:
Alright, merging to master / 1.6.
